packages/flutter/lib/src/material/navigation_rail.dart

```diff
@@ -0,0 +1,759 @@
+//            ],
+//            onDestinationSelected: (int index) {
+//              setState(() {
+//                _currentIndex = index;
+//              });
+//            },
+//          ),
+//          VerticalDivider(thickness: 1, width: 1),
+//          Expanded(
+//            child: Center(
+//              child: Text('currentIndex: $_currentIndex'),
+//            ),
+//          )
+//        ],
+//      ),
+//    );
+//  }
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [Scaffold], which can display the navigation rail within a [Row] of the
+///    [Scaffold.body] slot.
+///  * [NavigationRailDestination], which is used a model to create tappable
+///    destinations in the navigation rail.
+///  * [BottomNavigationBar], which is used as a horizontal alternative for
+///    the same style of navigation as the navigation rail.
+class NavigationRail extends StatefulWidget {
+  /// Creates a material design navigation rail.
+  ///
+  /// The argument [destinations] must not be null. Additionally, it must be
+  /// non-empty.
+  ///
+  /// If [elevation] is specified, it must be non-negative.
+  ///
+  /// If [preferredWidth] is specified, it must be non-negative, and if
+  /// [extendedWidth is specified, it must be non-negative and greater than
+  /// [preferredWidth].
+  ///
+  /// The argument [extended] must be non-null. [extended] can only be set to
+  /// true when when the [labelType] is null or [NavigationRailLabelType.none].
+  ///
+  /// If [backgroundColor], [elevation], [groupAlignment], [labelType],
+  /// [unselectedLabelTextStyle], [unselectedLabelTextStyle],
+  /// [unselectedIconTheme], or [selectedIconTheme] are null, then their
+  /// [NavigationRailThemeData] values will be used. If the corresponding
+  /// [NavigationRailThemeData] property is null, then the navigation rail
+  /// defaults are used.
+  ///
+  /// Typically used within a [Row] of the [Scaffold.body] property.
+  NavigationRail({
+    this.backgroundColor,
+    this.extended = false,
+    this.leading,
+    this.trailing,
+    @required this.destinations,
+    this.currentIndex = 0,
+    this.onDestinationSelected,
+    this.elevation,
+    this.groupAlignment,
+    this.labelType,
+    this.unselectedLabelTextStyle,
+    this.selectedLabelTextStyle,
+    this.unselectedIconTheme,
+    this.selectedIconTheme,
+    this.preferredWidth = _railWidth,
+    this.extendedWidth = _extendedRailWidth,
+  }) :  assert(destinations != null && destinations.isNotEmpty),
+        assert(0 <= currentIndex && currentIndex < destinations.length),
+        assert(elevation == null || elevation > 0),
+        assert(preferredWidth == null || preferredWidth > 0),
+        assert(extendedWidth == null || extendedWidth > 0),
+        assert((preferredWidth == null || extendedWidth == null) || extendedWidth >= preferredWidth),
+        assert(extended != null),
+        assert(!extended || (labelType == null || labelType == NavigationRailLabelType.none));
```

Comment:
Why is this a restriction? Could the `labelType` just be ignored if extended is true instead?

Reply:
Yeah it could be ignored but I think its better to disallow it (its a rare case that the labelType will be dynamic anyway)